430 add ability to search filter expenses

MilouH-TUE opened the pull request:
Closes #430 

I created this pull request in collaboration with @c5laungani . This merge request adds filter options for the bills list as requested in issue 430. We used the previous attempts for this issue as our starting point. Just before opening this pull request, we also observed that there was another issue (1155), that is a duplicate of the filtering issue. In a recent pull request for that issue, the feedback was to add tests. So in this pull request we tried to cover all feedback from previous attempts.

Screenshots:
![filter_closed](https://github.com/user-attachments/assets/654427c3-3452-4a15-9e5c-f806373ad093)
![filter_shown](https://github.com/user-attachments/assets/b6540b7e-aac1-4996-a783-481d41dd99e9)
![filter_applied](https://github.com/user-attachments/assets/d9272fa5-a63e-42f6-a41c-b1613bd725ec)

Demo video:
https://github.com/user-attachments/assets/40465943-3819-4158-b642-b3fc5899c065


Please let us know if there's anything we should look into further!

MilouH-TUE committed:
Implement filters in server side code

MilouH-TUE committed:
Store filters more efficiently

MilouH-TUE committed:
First attempt at tests

MilouH-TUE committed:
Make the tests work with the model

MilouH-TUE committed:
Clean up unnecessary whitespace differences

MilouH-TUE committed:
More whiteline reversions and re-adding old comment

MilouH-TUE committed:
Undo unused import

Jojo144 commented:
Thanks for your contribution! And sorry for the very late answer.
Here are my first comments:
- [ ] it should nice to have a "for whom" field, it is useful -> maybe too complicated in the UI for the usefulness
- [ ] we should be accent insensitive (useful in french) -> seems not so simple to do
- [ ] does "Clear filters" should really dismiss the panel? I would prefer not. -> maybe it's good like this, it does a real reset so in case of bug it is usefull
- [x] I think that the "Filter" button should be normal size instead of btn-sm

Jojo144 committed:
Bigger button

Jojo144 commented:
Also:
- [x] we should accept comma in "min" and "max" fields and replace it by a dot
- [x] the links are wrong when there are several pages (it clean the filters)

I tested on a project with lot of bills, it will be very handy!

Jojo144 committed:
Keep filters in pages when many bills

Jojo144 committed:
Make filters not disapearing when scrolling

Jojo144 committed:
Replace comma by dot in search amount

Jojo144 committed:
Indentation

Jojo144 commented:
For me this is ready for reviews

JocelynDelalande committed:
Improve factoring / re-use of query helpers

JocelynDelalande committed:
Improve UX around filter show/hide button

- More explicit labels on button, with relevant action verbs
- switch label between "show" and "hide" to reflect its action, trying to keep
  something stable for screen readers though
- add a opaque background to show/hide button (first plan was to add an opaque
  bg to the whole block, but it was making appearang weirder on the edges)

JocelynDelalande commented:
I tried to carefully test it in various cases… works great !

I fixed minor UX issues, see my commits.

I think that the top-bar over the list is kind-a clumsy, but this is clearly another topic. Same for accent-insensitive search, as I agree with @Jojo144 : seems non-trivial.

I'm excited we finally get filtering implemented ! Thanks @MilouH-TUE and @c5laungani to put it on the workbench again 🫶 ! And @Jojo144 for further testing and improving.

I hope this is the right one !

@Jojo144 I think you could squash all PR commits into one single commit (GH offers to do that with merge button), as our log is kinda-noisy :-)

Jojo144 committed:
Remove arrows

Jojo144 commented:
\o/
